pkg/federation/sdlmerge/sdlmerge.go

```diff
@@ -322,3 +325,66 @@ func (_ ScalarSharedType) valueRefs() []int {
 func (_ ScalarSharedType) valueName(_ int) string {
 	return ""
 }
+
+type PotentialEntityType interface {
+	getWalker() *astvisitor.Walker
+	getDocument() *ast.Document
+	assessValidEntity(ref int, nameBytes []byte) bool
+}
+
+func getPrimaryKeys(p PotentialEntityType, n *normalizer, name string, directiveRefs []int) map[string]bool {
+	baseKeys := n.entities[name]
+	document := p.getDocument()
+	primaryKeys := make(map[string]bool)
+	for _, directiveRef := range directiveRefs {
+		if document.DirectiveNameString(directiveRef) != "key" {
+			continue
+		}
+		directive := document.Directives[directiveRef]
+		if len(directive.Arguments.Refs) != 1 {
+			p.getWalker().StopWithExternalErr(operationreport.ErrKeyDirectiveMustHaveSingleArgument(name))
+		}
+		argumentRef := directive.Arguments.Refs[0]
+		if document.ArgumentNameString(argumentRef) != "fields" {
+			p.getWalker().StopWithExternalErr(operationreport.ErrKeyDirectiveMustHaveSingleArgument(name))
+		}
+		primaryKey := document.StringValueContentString(document.Arguments[argumentRef].Value.Ref)
+		if _, exists := baseKeys[primaryKey]; !exists || primaryKey == "" {
+			p.getWalker().StopWithExternalErr(operationreport.ErrPrimaryKeyReferencesMustExistOnEntity(primaryKey, name))
+		}
+		primaryKeys[primaryKey] = false
+	}
+	return primaryKeys
+}
+
+func checkAllPrimaryKeyReferencesAreExternal(p PotentialEntityType, name string, primaryKeys map[string]bool, fieldRefs []int) {
+	fieldReferences := len(primaryKeys)
+	if fieldReferences < 1 {
+		p.getWalker().StopWithExternalErr(operationreport.ErrEntityExtensionMustHaveKeyDirectiveAndExistingPrimaryKey(name))
+	}
+	document := p.getDocument()
+ParentLoop:
+	for _, fieldRef := range fieldRefs {
+		fieldName := document.FieldDefinitionNameString(fieldRef)
+		hasExternalDirective, isPrimaryKey := primaryKeys[fieldName]
+		if !isPrimaryKey {
+			continue
+		}
+		field := document.FieldDefinitions[fieldRef]
+		for _, directiveRef := range field.Directives.Refs {
+			if document.DirectiveNameString(directiveRef) != "external" {
+				continue
+			}
+			if !hasExternalDirective {
+				primaryKeys[fieldName] = true
+				fieldReferences -= 1
+			}
+			if fieldReferences == 0 {
+				return
+			}
+			continue ParentLoop
```

Comment:
yeah, I agree, don't try to use labels/goto in implementations

Reply:
I will try. This was just so it could return the error as soon as possible. And for my learning benefit, is there a reason why labels should be avoided?

Reply:
so more like bad practice, and there are very rare cases when you really need them
but some go std source code using them